[mlir][linalg] Implement Winograd Conv2D.

Hsiangkai opened the pull request:
This patch implements the Winograd Conv2D algorithm. It supports several configurations of Winograd Conv2D, including F(2, 3), F(4, 3) and F(2, 5). These configurations show that the implementation can support different kernel size (3 and 5) and different output size (2 and 4). Besides symetric kernel size 3x3 and 5x5, this patch also supports 1x3, 3x1, 1x5, and 5x1 kernels.

The implementation is based on the paper, Fast Algorithm for Convolutional Neural Networks. (https://arxiv.org/abs/1509.09308)

MaheshRavishankar reviewed (commented):
I'd like to get more input on this, but just for context, in IREE we handled this slightly differently. We implemented an operation for the input/filter/output transforms for winograd https://github.com/iree-org/iree/blob/9d60462ebe1bd5dcd0cc8cb8ca7a5e45523c1bd4/compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td#L983 . These implement the `TilingInterface` which allow executing these operations better (ability to distribute to threads, etc.) . They eventually get decomposed into what is being implemented here. This has worked well in IREE. I would suggest we try to "upstream" what is being done in IREE.

cc @Max191 @bjacob and @harsh-nod

MaheshRavishankar commented:
> > I'd like to get more input on this, but just for context, in IREE we handled this slightly differently. We implemented an operation for the input/filter/output transforms for winograd https://github.com/iree-org/iree/blob/9d60462ebe1bd5dcd0cc8cb8ca7a5e45523c1bd4/compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td#L983 . These implement the `TilingInterface` which allow executing these operations better (ability to distribute to threads, etc.) . They eventually get decomposed into what is being implemented here. This has worked well in IREE. I would suggest we try to "upstream" what is being done in IREE.
> > cc @Max191 @bjacob and @harsh-nod
> 
> Thanks for your information. I didn't dig into the implementation of IREE. I assume there exists some pass to tile the input into the size supported by Winograd Conv2D. (I might be wrong. I will figure it out.) I will take a look of TilingInterface, too. If you all agree to "upstream" IREE implementation to MLIR, I am fine to revoke this patch.

Yes, thats exactly what it does. We would like to upstream it, but not sure we have the bandwidth to do this. (Hence I havent marked this as change requested). We'd be happy to co-ordinate on upstreaming what is in IREE...

Hsiangkai commented:
I updated the patch to include `structured.winograd_conv2d` transform op and added a test case to show how to use `structured.tile_using_for` and `structured.winograd_conv2d` to deal with conv2d with arbitrary input size.

bjacob commented:
See also https://github.com/iree-org/iree/discussions/16571.

Hsiangkai commented:
We are willing to coordinate with you to upstream Winograd Conv2D algorithm. I have updated our implementation using TilingInterface. How about to use this version as a start point for the upstream work? If there is anything that I can do to align IREE’s implementation or meet IREE’s requirements, please tell me and I can help on it.

MaheshRavishankar reviewed (commented):
Cool! @Max191 can you help review this and shepherd this

ftynse commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgOps.td:
```diff
@@ -154,4 +154,96 @@ def Linalg_SoftmaxOp : Linalg_Op<"softmax",
+      "getTiledImplementation"]>]> {
+  let summary = "Winograd filter transform operator";
+  let description = [{
+    linalg.winograd_filter_transform transforms the filter of conv2D.
```
Please provide a more extended description, pseudo-code showing computation before/after or equivalent loop-level IR is welcome here.

ftynse commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgOps.td:
```diff
@@ -154,4 +154,96 @@ def Linalg_SoftmaxOp : Linalg_Op<"softmax",
+                       AnyRankedTensor:$output,
+                       I64Attr:$output_height,
+                       I64Attr:$output_width,
+                       I64Attr:$m,
+                       I64Attr:$r
```
Use more descriptive names.

Hsiangkai replied:
`m` and `r` are variable names in the Winograd algorithm. I added more description for the operators.

ftynse commented on mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h:
```diff
@@ -1312,6 +1312,20 @@ FailureOr<Operation *> transposeBatchMatmul(RewriterBase &rewriter,
+                                      linalg::Conv2DNhwcFhwcOp op, int64_t m,
+                                      int64_t r);
+FailureOr<Operation *>
+winogradConv2DRewriteFilterTransform(RewriterBase &rewriter,
```
Please document all top-level entities.

ftynse commented on mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h:
```diff
@@ -1692,6 +1706,11 @@ void populateTransposeMatmulPatterns(RewritePatternSet &patterns,
 void populateBlockPackMatmulPatterns(RewritePatternSet &patterns,
                                      const ControlBlockPackMatmulFn &controlFn);
 
+/// Patterns to apply Winograd Conv2D algorithm.
```
What's the difference between `2DPatterns` and `2DRewritePattersn`?

Hsiangkai replied:
I rename `2DRewritePatterns` to `DecomposeWinogradOpsPatterns`.

ftynse commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -2737,6 +2737,336 @@ FailureOr<SmallVector<Value>> SoftmaxOp::decomposeOperation(OpBuilder &b) {
+  auto heightAttr = builder.getI64IntegerAttr(getOutputHeight());
+  auto widthAttr = builder.getI64IntegerAttr(getOutputWidth());
+  Value output = getOutput();
+  for (auto dim = 0; dim < 4; ++dim) {
```
Expand `auto` into a proper type unless the RHS is a cast, or the type is too long to spell out.

ftynse commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -2737,6 +2737,336 @@ FailureOr<SmallVector<Value>> SoftmaxOp::decomposeOperation(OpBuilder &b) {
+  Location loc = getLoc();
+  Value zero = builder.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = builder.create<arith::ConstantIndexOp>(loc, 1);
+  auto heightAttr = builder.getI64IntegerAttr(getOutputHeight());
+  auto widthAttr = builder.getI64IntegerAttr(getOutputWidth());
```
`getOutputHeightAttr()` will return the attribute without the need to reconstruct it from value locally.

ftynse commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -2737,6 +2737,336 @@ FailureOr<SmallVector<Value>> SoftmaxOp::decomposeOperation(OpBuilder &b) {
+    loopBounds[dim].size = getDimValue(builder, loc, output, dim);
+    loopBounds[dim].stride = one;
+  }
+  // Iterate on output domain
```
```suggestion
  // Iterate on output domain.
```

Nit: always terminate sentences in comments with a full stop.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/WinogradConv2D.cpp:
```diff
@@ -0,0 +1,1022 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+/// \file
```
We don't do this in most of the code.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/WinogradConv2D.cpp:
```diff
@@ -0,0 +1,1022 @@
+  return insertSliceOp;
+}
+
+Value collaps2DData(RewriterBase &rewriter, Location loc, Value data) {
```
collapse

ftynse reviewed (changes requested):
Thank you for your contribution! This is a rather big change so I would strongly suggest splitting it into several parts if you don't want to be blocked waiting on reviewers to find time to get through 2k LoC. For example:
- introduce the data modification operations + tests;
- implement the tiling interface for them + tests;
- implement the transformation and one way of executing it (test pass or transform dialect);
- implement the other way of executing it.

This will let reviewers focus on relevant aspects and not require committing a big block of time.

I left a couple of stylistic comments. The big one is to please document all functions and operations extensively.

If there is a design discussion to be had, please do so on https://discourse.llvm.org/c/mlir/31 rather than in PR comments. In particular, choosing the design for upstream MLIR is not conditioned on the needs of a particular downstream, however large.

Max191 commented on mlir/lib/Dialect/Linalg/Transforms/WinogradConv2D.cpp:
```diff
@@ -0,0 +1,1022 @@
+  Type elementType = filterType.getElementType();
+  int64_t alphaH = leftTransform ? m + r - 1 : 1;
+  int64_t alphaW = rightTransform ? m + r - 1 : 1;
+  int64_t retHeight = leftTransform ? (outputH / m) * alphaH : 1;
+  int64_t retWidth = rightTransform ? (outputW / m) * alphaW : 1;
```
It looks like the output shape may be too small to capture the full input when `m` does not evenly divide the output shape. You could use a ceildiv here instead, and then pad the input when the winograd op gets decomposed. This way you can use the last partial tile of the input, and capture the full input tensor.

Hsiangkai replied:
Updated the patch to consider unaligned cases and added tests for unaligned cases.

Max191 reviewed (changes requested):
Thanks for working on adding these ops upstream, and with more options on kernel size/input tile size than what we have in IREE! Just some higher level op semantics comments for now.

The main difference I see between this implementation and the IREE implementation is that the dimensionality of the transforms are different. In this implementation, the 4D input tensor is transformed into another 4D transformed tensor, while in the IREE implementation, a 6D input tensor is produced. The extra dimensionality comes from having `(alpha)x(alpha)` dimensions expanded from the `H` and `W` dimensions. In other words, with an `NHWC` layout, the IREE input transform would produce `tensor<TxTxNxceil(H/T)xceil(W/T)xC>`.

This extra dimensionality is very useful for tiling, since the iteration space is actually over the innermost of those 6 dimensions. Trying to implement a tiled implementation where some of those dimensions are collapsed is tricky.

I would suggest reworking the op semantics a bit to allow for this extra dimensionality. Then, the tiled implementation from IREE should also just drop right in here. Also, another benefit of having separate input tile dimensions is that the input tile dimensions can be innermost if desired. This can potentially provide performance benefits from having continuous accesses in the winograd ops. We have experimentally tried this in IREE, but have not merged the change yet.

The other difference I notice, is that there is no implicit padding on the input transform. This means that for shapes that are unaligned with the input tiles, the input transform will not capture the last partial tile of the input tensor. Instead the input transform should extract the last partial tile, and pad the slice with zeros to the input tile size (alpha).

The last comment is that the implementation here has explicit I64Attrs for the output height and width. This should not be necessary if the ops have padding semantics as suggested above, since there is no need to know the output shape for the input or filter transform in that case. Having this attribute means that the op is restricted to static image sizes only (making this an operand could allow some dynamic support, but it is better not to rely on having these sizes since it is not necessary).

Hsiangkai committed:
[mlir][linalg] Implement Conv2D using Winograd Conv2D algorithm

Define high level winograd operators and convert conv_2d_nhwc_fhwc into
winograd operators. According to Winograd Conv2D algorithm, we need
three transform operators for input, filter, and output transformation.

The formula of Winograd Conv2D algorithm is

Y = A^T x [(G x g x G^T) @ (B^T x d x B)] x A

filter transform: G x g x G^T
input transform: B^T x d x B
output transform: A^T x y x A

The implementation is based on the paper, Fast Algorithm for
Convolutional Neural Networks. (https://arxiv.org/abs/1509.09308)

Hsiangkai committed:
[mlir][linalg] Implement TilingInterface for winograd operators

In order to support arbitrary size input data of conv2d, implement
TilingInterface for winograd operators. Before converting winograd
operators into nested loops with matrix multiply, tile the input of
conv2d into the supported size first.

Add a transform operator structured.decompose_winograd_op to decompose
winograd operators. Before applying the transform op, use tile_using_for
to tile the input data into supported size. The test case shows how to
tile and decompose winograd operators.

Hsiangkai committed:
[mlir][linalg] Add transform operator for Winograd Conv2D algorithm

Add a transform operator structured.winograd_conv2d to convert
linalg.conv_2d_nhwc_fhwc to Linalg winograd operators.

Hsiangkai committed:
[mlir][linalg] Decompose winograd operators

Convert Linalg winograd_filter_transform, winograd_input_transform, and
winograd_output_transform into nested loops with matrix multiplication
with constant transform matrices.

Support several configurations of Winograd Conv2D, including F(2, 3),
F(4, 3) and F(2, 5). These configurations show that the implementation
can support different kernel size (3 and 5) and different output size
(2 and 4). Besides symetric kernel size 3x3 and 5x5, this patch also
supports 1x3, 3x1, 1x5, and 5x1 kernels.

The implementation is based on the paper, Fast Algorithm for
Convolutional Neural Networks. (https://arxiv.org/abs/1509.09308)

Hsiangkai commented:
> Thank you for your contribution! This is a rather big change so I would strongly suggest splitting it into several parts if you don't want to be blocked waiting on reviewers to find time to get through 2k LoC. For example:
> 
> * introduce the data modification operations + tests;
> * implement the tiling interface for them + tests;
> * implement the transformation and one way of executing it (test pass or transform dialect);
> * implement the other way of executing it.
> 
> This will let reviewers focus on relevant aspects and not require committing a big block of time.
> 
> I left a couple of stylistic comments. The big one is to please document all functions and operations extensively.
> 
> If there is a design discussion to be had, please do so on https://discourse.llvm.org/c/mlir/31 rather than in PR comments. In particular, choosing the design for upstream MLIR is not conditioned on the needs of a particular downstream, however large.

Thanks for your review. I split the original patch into 4 smaller commits and address all your comments.

Hsiangkai commented:
> Thanks for working on adding these ops upstream, and with more options on kernel size/input tile size than what we have in IREE! Just some higher level op semantics comments for now.
> 
> The main difference I see between this implementation and the IREE implementation is that the dimensionality of the transforms are different. In this implementation, the 4D input tensor is transformed into another 4D transformed tensor, while in the IREE implementation, a 6D input tensor is produced. The extra dimensionality comes from having `(alpha)x(alpha)` dimensions expanded from the `H` and `W` dimensions. In other words, with an `NHWC` layout, the IREE input transform would produce `tensor<TxTxNxceil(H/T)xceil(W/T)xC>`.
> 
> This extra dimensionality is very useful for tiling, since the iteration space is actually over the innermost of those 6 dimensions. Trying to implement a tiled implementation where some of those dimensions are collapsed is tricky.
> 
> I would suggest reworking the op semantics a bit to allow for this extra dimensionality. Then, the tiled implementation from IREE should also just drop right in here. Also, another benefit of having separate input tile dimensions is that the input tile dimensions can be innermost if desired. This can potentially provide performance benefits from having continuous accesses in the winograd ops. We have experimentally tried this in IREE, but have not merged the change yet.
> 
> The other difference I notice, is that there is no implicit padding on the input transform. This means that for shapes that are unaligned with the input tiles, the input transform will not capture the last partial tile of the input tensor. Instead the input transform should extract the last partial tile, and pad the slice with zeros to the input tile size (alpha).
> 
> The last comment is that the implementation here has explicit I64Attrs for the output height and width. This should not be necessary if the ops have padding semantics as suggested above, since there is no need to know the output shape for the input or filter transform in that case. Having this attribute means that the op is restricted to static image sizes only (making this an operand could allow some dynamic support, but it is better not to rely on having these sizes since it is not necessary).

Thanks for your review and suggestions. It's very helpful to get insight of IREE implementation. I appreciate your detail explanation. I will address your comments in the next update.

Hsiangkai commented:
Update the patch using 6D tensors after filter/input transformation, instead of 4D.

Hsiangkai commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -2737,6 +2737,336 @@ FailureOr<SmallVector<Value>> SoftmaxOp::decomposeOperation(OpBuilder &b) {
+  sliceSizes.push_back(sizes[2]);
+  sliceSizes.push_back(alphaHAttr);
+  sliceSizes.push_back(alphaWAttr);
+  sliceSizes.push_back(sizes[3]);
```
I updated my patch to use the number of tiles as additional dimensions.

Hsiangkai commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgOps.td:
```diff
@@ -154,4 +154,96 @@ def Linalg_SoftmaxOp : Linalg_Op<"softmax",
   let hasVerifier = 1;
 }
 
+def Linalg_WinogradFilterTransformOp : Linalg_Op<"winograd_filter_transform",
```
Done.

Hsiangkai commented:
I think I already addressed all the comments. I split the original patch into 4 commits. I am not sure I do it in the right way or not. Should I create 4 separate PR instead of 1 PR with 4 commits? Please help me to review it again. Thank you.

Hsiangkai commented:
I created four PRs for Winograd Conv2D implementation.

https://github.com/llvm/llvm-project/pull/96176
https://github.com/llvm/llvm-project/pull/96177
https://github.com/llvm/llvm-project/pull/96178
https://github.com/llvm/llvm-project/pull/96179

Let's move the review there.
